tests/bcl-test/templates/common/TestRunner.Core/TcpTextWriter.cs

```diff
@@ -62,25 +64,41 @@ static string SelectHostName (string[] names, int port)
 			return result;
 		}
 
-		public TcpTextWriter (string hostName, int port)
+		public TcpTextWriter (string hostName, int port, bool isTunnel = true)
 		{
 			if ((port < 0) || (port > ushort.MaxValue))
 				throw new ArgumentOutOfRangeException (nameof (port), $"Port must be between 0 and {ushort.MaxValue}" );
 
 			if (hostName == null)
 				throw new ArgumentNullException (nameof (hostName));
-			HostName = SelectHostName (hostName.Split (','), port);
+			if (!isTunnel)
+				HostName = SelectHostName (hostName.Split (','), port);
+
 			Port = port;
 
 #if __IOS__
 			UIApplication.SharedApplication.NetworkActivityIndicatorVisible = true;
 #endif
 
 			try {
-				client = new TcpClient (HostName, port);
+				if (isTunnel) {
+					server = new TcpListener (IPAddress.Any, Port);
+					server.Start ();
+					client = server.AcceptTcpClient ();
+					// block until we have the ping from the client side
```

Comment:
block until... what's the caller/harness timeout ?
do we want to wait that long ?
will it report something useful to diagnose ?

Reply:
hmm you make me use my brain too much... if harness dies, we will have issues. We could do:
```csharp
server = new TcpListener (IPAddress.Any, Port);
server.Server.ReceiveTimeout = 5000;
```
That sets the socket receive timeout in milliseconds.